### PR TITLE
Port to new pymbolic

### PR DESCRIPTION
**Pull request by alexfikl**
Updates things to use `expr_dataclass` and fixes various deprecation warnings.

Needs:
- [X] https://github.com/inducer/pymbolic/pull/150
- [X] https://github.com/inducer/sumpy/pull/217
- [X] https://github.com/inducer/loopy/pull/868

### CODE REVIEW

**Comment by alexfikl** [comment]
This seems to be in a better state than expected! The remaining deprecation warnings are from `loopy` and `sumpy` and it just need a little change in `pymbolic` for `__eq__` in `IntG` to work.

**Comment by inducer** [comment]
I have some changes based on https://github.com/inducer/pymbolic/pull/159 that I'd like to push. I imagine you have some changes based on https://github.com/inducer/pymbolic/pull/160 that you'd like to push. How do we sort them to avoid stomping on one another's feet?

**Comment by inducer** [comment]
> This will start complaining once [inducer/pymbolic#160](https://github.com/inducer/pymbolic/pull/160) is merged though 😁

Uh-oh. Better get this in quickly. :)

File: pytential/symbolic/dof_desc.py [hunk 1]
```diff
@@ -275,6 +275,8 @@ def as_dofdesc(desc: DOFDescriptorLike) -> DOFDescriptor:
 
 # {{{ type annotations
 
+DEFAULT_DOFDESC = DOFDescriptor()
```
Comment (inducer) [inline pytential/symbolic/dof_desc.py:278]: Renamed this guy `DEFAULT_DOFDESC`, hope that's OK.

Comment (alexfikl) [reply pytential/symbolic/dof_desc.py:278]: Yeah, definitely better! Not sure why I named it empty..

**Pull request merged**